### PR TITLE
Let IRgen compile predicate IR modules specifically.

### LINKED ISSUE #1655: Annotate predicate IR with `predicate` and run the appropriate `asm` checks
https://github.com/FuelLabs/sway/pull/1640 enables predicates to compile just like scripts which is perfectly fine from a functional point of view. However, if we want some predicate-specific `asm` checks to run in codegen, then we need to indicate, in the IR and beyond, that we're actually compiling a script not a predicate. This can be accomplished by making sure that the IR corresponding to predicates always start with `predicate`, just like we do for scripts and contracts. That is:
```rust
predicate {
    fn main() -> bool {
        ...
    }
    ...
}
```
Then, use this information to run whatever codegen checks we want to run (purity checks and whatnot)

### PR DESCRIPTION
**Pull request by otrho**
Previously we were just compiling scripts and predicates as the same and so predicates were compiled to `script` modules.

Closes #1655.

### CODE REVIEW

**Comment by otrho** [comment]
#1655 talks about ASM checks but they're managed in #2550, which was basically blocked by this.

**Comment by otrho** [comment]
Hrmm, shoulda probably run the tests myself instead of just :crossed_fingers: and :pray:.

File: sway-core/src/asm_generation/from_ir.rs [hunk 1]
```diff
@@ -115,7 +115,8 @@ fn compile_module_to_asm(
     let kind = match module.get_kind(context) {
         Kind::Contract => ProgramKind::Contract,
         Kind::Script => ProgramKind::Script,
-        Kind::Library | Kind::Predicate => todo!("libraries and predicates coming soon!"),
+        Kind::Predicate => ProgramKind::Predicate,
+        Kind::Library => todo!("libraries and predicates coming soon!"),
```
Comment (mohammadfawaz) [inline sway-core/src/asm_generation/from_ir.rs:119]:
```suggestion
        Kind::Library => todo!("libraries coming soon!"),
```

**Review by mohammadfawaz — APPROVED** [review]

**Comment by mohammadfawaz** [comment]
> instead of just 🤞 and 🙏

I thought that's what CI is for 😆

**Pull request merged**